web_services/foxx/brightid/operations.js

```diff
@@ -209,24 +231,50 @@ function getMessage(op) {
   return stringify(signedOp);
 }
 
-function updateSponsorOp(op) {
-  const { sponsorPrivateKey, context } = db.getApp(op.app);
+function sponsor(app, contextId) {
+  const { sponsorPrivateKey, context } = db.getApp(app);
+
+  if (!sponsorPrivateKey || !db.getContext(context)) {
+    throw 'can not relay sponsor requests for this app';
+  } else if (db.unusedSponsorships(app) < 1) {
+    throw 'app does not have unused sponsorships';
+  }
+
   const { collection, idsAsHex } = db.getContext(context);
   const coll = arango._collection(collection);
 
   if (idsAsHex) {
-    op.contextId = op.contextId.toLowerCase();
+    contextId = contextId.toLowerCase();
   }
 
-  op.id = db.getUserByContextId(coll, op.contextId)
-  if (!op.id) {
-    throw 'unlinked context id';
-  }
+  const sponsorshipsColl = arango._collection('sponsorships');
+  const id = db.getUserByContextId(coll, contextId);
 
-  delete op.contextId;
-  let message = getMessage(op);
-  op.sig = uInt8ArrayToB64(Object.values(nacl.sign.detached(strToUint8Array(message), b64ToUint8Array(sponsorPrivateKey))));
-  op.hash = hash(message);
+  if (id) {
+    // contextId is linked to a brightid
+    const timestamp = Date.now();
+    db.sponsor(id, app, timestamp);
+    op = { name: 'Sponsor', app, id, timestamp, v: 5 }
+    let message = getMessage(op);
+    op.sig = uInt8ArrayToB64(Object.values(nacl.sign.detached(strToUint8Array(message), b64ToUint8Array(sponsorPrivateKey))));
+    op.hash = hash(message);
+    op.state = 'init';
+    db.upsertOperation(op);
+    // remove temporary sponsorship if exists
+    sponsorshipsColl.removeByExample({
+      _from: 'users/0',
+      contextId: contextId
+    });
+  } else {
+    // contextId is not linked to a brightid yet
+    // add a temporary sponsorship to be applied after user linked contextId
+    sponsorshipsColl.insert({
+      _from: 'users/0',
+      _to: 'apps/' + app,
+      expireDate: Math.ceil((Date.now() / 1000) + 3600),
```

Comment:
when is the expiration supposed to be?  10 hours in the future?  If you aren't going to use the date math functions, at least put a comment explaining what you expect the result to be.

Reply:
It will expire after one hour.
https://github.com/BrightID/BrightID-Node/commit/0b1539c3cbcadc3038d0c73866a52f38feb500e7